api/deploy.go

```diff
@@ -86,44 +86,59 @@ func newServiceDeployer(api *API, env map[string]string, options ...DeployServic
 
 // FromURL deploys a service from git or tarball url.
 func (d *serviceDeployer) FromURL(url string) (*service.Service, *importer.ValidationError, error) {
+	path, err := d.createTempDir()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer os.RemoveAll(path)
+	d.sendStatus("Downloading service...", Running)
 	defer d.closeStatus()
 	if xgit.IsGitURL(url) {
-		d.sendStatus("Downloading service...", Running)
-		path, err := d.createTempDir()
-		if err != nil {
-			return nil, nil, err
-		}
 		defer os.RemoveAll(path)
 		if err := xgit.Clone(url, path); err != nil {
 			return nil, nil, err
 		}
 
-		// XXX: remove .git folder from repo.
-		// It makes docker build iamge id same between repo clones.
-		if err := os.RemoveAll(filepath.Join(path, ".git")); err != nil {
+	} else {
+		// if not git repo then it should be tarball
+		resp, err := http.Get(url)
+		if err != nil {
 			return nil, nil, err
 		}
-
-		d.sendStatus("Service downloaded with success", DonePositive)
-		r, err := xarchive.GzippedTar(path, nil)
-		if err != nil {
+		defer resp.Body.Close()
+		if err := archive.Untar(resp.Body, path, nil); err != nil {
 			return nil, nil, err
 		}
-		defer r.Close()
-		return d.deploy(r)
 	}
 
-	// if not git repo then it should be tarball
-	resp, err := http.Get(url)
+	// XXX: remove .git folder from repo.
+	// It makes docker build iamge id same between repo clones.
+	if err := os.RemoveAll(filepath.Join(path, ".git")); err != nil {
+		return nil, nil, err
+	}
+
+	// NOTE: this is check for tar repos, if there is only one
+	// directory inside untar archive set temp path to it.
+	dirs, err := ioutil.ReadDir(path)
 	if err != nil {
 		return nil, nil, err
 	}
-	defer resp.Body.Close()
-	return d.deploy(resp.Body)
+	if len(dirs) == 1 && dirs[0].IsDir() {
+		path = filepath.Join(path, dirs[0].Name())
+	}
+
+	d.sendStatus("Service downloaded with success", DonePositive)
+	r, err := archive.Tar(path, archive.Uncompressed)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer r.Close()
```

Comment:
This is the only place where a ReadCloser from a tarball is close.
Either we should close ALL of them, either we should not close.
I guess we should close ALL of them?

Eg: `decompressedStream` in `service/service.go`